dev_tools/profiling/quantum_volume.py

```diff
@@ -0,0 +1,144 @@
+# limitations under the License.
+"""Tool to run the Quantum Volume benchmark defined by IBM in
+https://arxiv.org/abs/1811.12926."""
+
+import argparse
+from collections import defaultdict
+import math
+import sys
+from typing import DefaultDict, List
+
+import numpy as np
+
+import cirq
+
+
+def generate_model_circuit(num_qubits: int, depth: int) -> cirq.Circuit:
+    """Generates a model circuit with the given number of qubits and depth.
+
+    The generated circuit consists of `depth` layers of random qubit
+    permutations followed by random two-qubit gates that are sampled from the
+    Haar measure on SU(r).
+
+    Args:
+        num_qubits: The number of qubits in the generated circuit.
+        depth: The number of layers in the circuit.
+
+    Returns:
+      The generated circuit.
+    """
+    # Setup the circuit and its qubits.
+    qubits = [cirq.GridQubit(0, i) for i in range(num_qubits)]
+    circuit = cirq.Circuit()
+
+    # For each layer.
+    for _ in range(depth):
+        # Generate uniformly random permutation Pj of [0...n-1]
+        perm = np.random.permutation(num_qubits)
+
+        # For each consecutive pair in Pj, generate Haar random SU(4)
+        # Decompose each SU(4) into CNOT + SU(2) and add to Ci
+        for k in range(math.floor(num_qubits / 2)):
+            permuted_qubits = [int(perm[2 * k]), int(perm[2 * k + 1])]
+            special_unitary = cirq.testing.random_special_unitary(4)
+            kak_unitary = cirq.unitary(cirq.kak_decomposition(special_unitary))
+
+            # Convert the decomposed unitary to Cirq operations and add them to
+            # the circuit.
+            ops = cirq.two_qubit_matrix_to_operations(
+                qubits[permuted_qubits[0]], qubits[permuted_qubits[1]],
+                kak_unitary, False)
+            circuit.append(ops)
+
+    # Measure all of the qubits at the end of the circuit.
+    circuit.append([cirq.measure(qubit) for qubit in qubits])
+    return circuit
+
+
+def compute_heavy_set(circuit: cirq.Circuit, seed: int = None) -> List[str]:
+    """Classically compute the heavy set of the given circuit.
+
+    The heavy set is defined as the output bit-strings that have a greater than
+    median probability of being generated.
+
+    Args:
+        circuit: The circuit to classically simulate.
+        seed: A seed to pass to the simulator.
+
+    Returns:
+        A list containing all of the heavy bit-string results.
+    """
+    # Run the simulation 100 times, storing the a dict from resulting bit-string
+    # to the number of times that bit-string came up.
+    results: DefaultDict[str, int] = defaultdict(int)
+    for _ in range(100):
+        simulator = cirq.Simulator(seed=seed)
+        result = simulator.run(circuit)
+        bit_string = "".join(str(i) for i in result.data.iloc[0])
+        results[bit_string] += 1
+
+    # Compute the median probability of the output bit-strings.
+    median = np.median(list(results.values()))
```

Comment:
This is not the median you want. 

Calculating hog requires one call to `simulator.simulate()` (which can only be done classically with exponential cost). This will give you a vector of 2**n_qubits probabilities, of which you can find the median. 

*Then* you can call `sampler.run()` for a number of repetitions. For each bitstring you sample, you can look up its true probability and figure out if it's higher than the true median

Reply:
Ah of course, thanks for catching that. Done, PTAL.